apiserver/facades/client/application/application_unit_test.go

```diff
@@ -1126,6 +1131,272 @@ func (s *ApplicationSuite) TestDeployCharmOrigin(c *gc.C) {
+}
+
+func (s *ApplicationSuite) TestApplicationDeployWithPlacementLockedError(c *gc.C) {
+	ctrl := s.setup(c)
+	defer ctrl.Finish()
+
+	machine := s.expectMachine(ctrl)
+	machine.EXPECT().IsLockedForSeriesUpgrade().Return(true, nil).MinTimes(1)
+	s.backend.EXPECT().Machine("0").Return(machine, nil).MinTimes(1)
+	containerMachine := s.expectMachine(ctrl)
+	containerMachine.EXPECT().IsLockedForSeriesUpgrade().Return(false, nil).MinTimes(1)
+	containerMachine.EXPECT().IsParentLockedForSeriesUpgrade().Return(true, nil).MinTimes(1)
+	s.backend.EXPECT().Machine("0/lxd/0").Return(containerMachine, nil).MinTimes(1)
+
+	curl := charm.MustParseURL("cs:precise/dummy-42")
+	args := []params.ApplicationDeploy{{
+		ApplicationName: "machine-placement",
+		CharmURL:        curl.String(),
+		Placement:       []*instance.Placement{{Scope: "#", Directive: "0"}},
+	}, {
+		ApplicationName: "container-placement",
+		CharmURL:        curl.String(),
+		Placement:       []*instance.Placement{{Scope: "lxd", Directive: "0"}},
+	}, {
+		ApplicationName: "container-placement-locked-parent",
+		CharmURL:        curl.String(),
+		Placement:       []*instance.Placement{{Scope: "#", Directive: "0/lxd/0"}},
+	}}
+	results, err := s.api.Deploy(params.ApplicationsDeploy{
+		Applications: args,
+	})
+
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results, gc.HasLen, 3)
+	c.Assert(results.Results[0].Error.Error(), gc.Matches, ".*: machine is locked for series upgrade")
+	c.Assert(results.Results[1].Error.Error(), gc.Matches, ".*: machine is locked for series upgrade")
+	c.Assert(results.Results[2].Error.Error(), gc.Matches, ".*: parent machine is locked for series upgrade")
+}
+
+func (s *ApplicationSuite) TestApplicationDeployFailCharmOrigin(c *gc.C) {
+	defer s.setup(c).Finish()
+
+	originId := createCharmOriginFromURL(c, charm.MustParseURL("cs:jammy/test-42"))
+	originId.ID = "testingID"
+	originHash := createCharmOriginFromURL(c, charm.MustParseURL("cs:jammy/test-42"))
+	originHash.Hash = "testing-hash"
+
+	args := []params.ApplicationDeploy{{
+		CharmOrigin: originId,
+	}, {
+		CharmOrigin: originHash,
+	}}
+	results, err := s.api.Deploy(params.ApplicationsDeploy{
+		Applications: args,
+	})
+
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results, gc.HasLen, 2)
+	c.Assert(results.Results[0].Error.Code, gc.Equals, "bad request")
+	c.Assert(results.Results[1].Error.Code, gc.Equals, "bad request")
+}
+
+func (s *ApplicationSuite) TestApplicationDeploymentRemovesPendingResourcesOnFailure(c *gc.C) {
+	ctrl := s.setup(c)
+	defer ctrl.Finish()
+
+	resources := map[string]string{"dummy": "pending-id"}
+	resourcesManager := mocks.NewMockResources(ctrl)
+	resourcesManager.EXPECT().RemovePendingAppResources("my-app", resources).Return(nil)
+	s.backend.EXPECT().Resources().Return(resourcesManager)
+
+	results, err := s.api.Deploy(params.ApplicationsDeploy{
+		Applications: []params.ApplicationDeploy{{
+			// Missing a charmURL, so this will error on deployApplication
```

Comment:
Need to check for the error if it's expected.

Reply:
Ah I see I haven't explained this particularly clearly. This took me a while to understand my intention

This ApplicationDeploy params object is constructed in a way to intentionally error out of deployApplication early, to test that in the case of a failed deployment pending resources are removed

I will Improve the comment to better reflect this